File: solutions/3075. Maximize Happiness of Selected Children/3075.cpp
// Time complexity: O(nlogn)
// Space complexity: O(1)
class Solution {
public:
  long long maximumHappinessSum(vector<int>& happiness, int k) {
    sort(happiness.begin(), happiness.end(), greater<int>());
    long long count = 0;
    long long sum = 0;
        
    for (int i = 0; i < k; ++i) {
      if (happiness[i] - count < 0)
        return sum;

      sum += happiness[i] - count;
      ++count;
    }
        
    return sum;
  }
};

File: solutions/3075. Maximize Happiness of Selected Children/3075.rs
// Time complexity: O(nlogn)
// Space complexity: O(n)
use std::cmp::max;

impl Solution {
  pub fn maximum_happiness_sum(mut happiness: Vec<i32>, k: i32) -> i64 {
    happiness.sort();
    happiness.reverse();
    *&happiness[0..k as usize].iter()
    .enumerate()
    .map(|(i, x)| max(0, *x as i64 - i as i64)).sum()
  }
}